### PR TITLE
IOS interface model

### PR DESCRIPTION
**Pull request by justjais**
**Docstring**
```
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright 2019 Red Hat Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

##############################################
################# WARNING ####################
##############################################
###
### This file is auto generated by the resource
###   module builder playbook.
###
### Do not edit this file manually.
###
### Changes to this file will be over written
###   by the resource module builder.
###
### Changes should be made in the model used to
###   generate this file or in the resource module
###   builder template.
###
##############################################
##############################################
##############################################

"""
The module file for ios_interfaces
"""

from __future__ import absolute_import, division, print_function
from ansible.module_utils.basic import AnsibleModule
from ansible_collections.ansible_network.network.plugins.module_utils. \
     ios.config.interfaces.interfaces import Interfaces

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': [u'preview'],
                    'supported_by': [u'Ansible Network']}


DOCUMENTATION = """
---
module: ios_interfaces
version_added: 2.9
short_description: Manage Interface on Cisco IOS network devices.
description: Manage Interface on Cisco IOS network devices.
author: [u'Sumit Jaiswal (@justjais)']
options:
  config:
    description: The provided configuration
    suboptions:
      name:
        description:
        - Name of interface, i.e. GigabitEthernet0/2, loopback999.
        type: str
        required: true
      description:
        description:
        - Description of Interface.
        type: str
      enabled:
        description:
        - Interface link status.
        type: bool
        default: true
      speed:
        description:
        - Interface link speed. Applicable for ethernet interface only.
        type: str
      mtu:
        description:
        - Maximum size of transmit packet. Must be an number between 64 and 9600.
          Applicable for Ethernet interface only.
        type: str
      duplex:
        description:
        - Interface link status. Applicable for ethernet interface only.
        type: str
        default: auto
        choices: ['full', 'half', 'auto']
  state:
    choices:
    - merged
    - replaced
    - overridden
    - deleted
    default: merged
    description:
    - The state the configuration should be left in
    type: str
"""

EXAMPLES = """
---

# Using merged

- name: Configure Ethernet interfaces
  ios_interfaces:
    config:
      - name: GigabitEthernet0/2
        description: 'Configured by Ansible'
        enabled: True
      - name: GigabitEthernet0/3
        description: 'Configured by Ansible Network'
        enabled: False
        duplex: full
    operation: merged

# Using replaced

- name: Configure following interfaces and replace their existing config
  ios_interfaces:
    config:
      - name: GigabitEthernet0/2
        description: Configured by Ansible
        enabled: True
        mtu: 2000
      - name: GigabitEthernet0/3
        description: 'Configured by Ansible Network'
        enabled: False
        duplex: auto
    operation: replaced

# Using overridden

- name: Override interfaces
  ios_interfaces:
    config:
      - name: GigabitEthernet0/2
        description: 'Configured by Ansible'
        enabled: True
        duplex: auto
      - name: GigabitEthernet0/3
        description: 'Configured by Ansible Network'
        enabled: False
        speed: 1000
    operation: overridden

# Using deleted

- name: Delete IOS interfaces as in given arguments
  ios_interfaces:
    config:
      - name: GigabitEthernet0/2
        description: 'Configured by Ansible'
      - name: GigabitEthernet0/3
        description: 'Configured by Ansible Network'
        mtu: 1800
    operation: deleted

"""
```
**Layout**
```
plugins
   ├── action
   ├── filter
   ├── inventory
   ├── module_utils
   │   ├── __init__.py
   │   ├── ios
   │   │   ├── __init__.py
   │   │   ├── argspec
   │   │   │   ├── __init__.py
   │   │   │   ├── facts
   │   │   │   │   ├── __init__.py
   │   │   │   │   └── facts.py
   │   │   │   └── interfaces
   │   │   │       ├── __init__.py
   │   │   │       └── interfaces.py
   │   │   ├── config
   │   │   │   ├── __init__.py
   │   │   │   ├── base.py
   │   │   │   └── interfaces
   │   │   │       ├── __init__.py
   │   │   │       └── interfaces.py
   │   │   ├── facts
   │   │   │   ├── __init__.py
   │   │   │   ├── base.py
   │   │   │   ├── facts.py
   │   │   │   └── interfaces
   │   │   │       ├── __init__.py
   │   │   │       └── interfaces.py
   │   │   └── utils
   │   │       ├── __init__.py
   │   │       └── utils.py
   │   └── network
   │       ├── __init__.py
   │       └── argspec
   │           ├── __init__.py
   │           └── base.py
   └── modules
       ├── __init__.py
       ├── ios_facts.py
       └── ios_interfaces.py
```

### CODE REVIEW

**Comment by samccann** [comment]
Please also add the before and after configuration snippets to the examples. See the eos_interfaces model for how this was handled - https://github.com/ansible-network/resource_module_models/pull/7/files

File: models/ios/interfaces/merged_example_01.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Using merged
+
+- name: Configure Ethernet interfaces
```
Comment (samccann) [inline models/ios/interfaces/merged_example_01.txt]:
```suggestion
- name: Merge provided configuration with device configuration
```

File: models/ios/interfaces/override_example_01.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+
+# Using overridden
+
+- name: Override interfaces
```
Comment (samccann) [inline models/ios/interfaces/override_example_01.txt]:
```suggestion
- name: Override device configuration of all interfaces with provided configuration
```

File: models/ios/interfaces/replaced_example_01.txt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Using replaced
+
+- name: Configure following interfaces and replace their existing config
```
Comment (samccann) [inline models/ios/interfaces/replaced_example_01.txt]:
```suggestion
- name: Replaces device configuration of listed interfaces with provided configuration
```

File: models/ios/interfaces/ios_interface.yaml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        - Description of Interface.
+        type: str
+      enabled:
+        description:
+        - Interface link status.
+        type: bool
+        default: true
+      speed:
+        description:
+        - Interface link speed. Applicable for ethernet interface only.
```
[the remark this answers is not in the document]
Comment (justjais) [reply models/ios/interfaces/ios_interface.yaml]: IOS speed has the following supported value: 10, 100, 1000, auto

File: models/ios/interfaces/deleted_example_01.txt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#  speed 1000
+# interface GigabitEthernet0/3
+#  description Configured and Replaced by Ansible Network
+#  mtu 2500
+#  no ip address
+#  shutdown
+#  duplex full
+#  speed 1000
+
+- name: Delete IOS interfaces as in given arguments
```
Comment (samccann) [inline models/ios/interfaces/deleted_example_01.txt]:
```suggestion
- name: Delete attributes of given interfaces (Note: This won't delete the interface itself)
```

**Pull request merged**